google/cloud/pubsub/internal/tracing_message_callback_test.cc

```diff
@@ -46,42 +47,85 @@ using ::google::cloud::testing_util::SpanKindIsInternal;
 using ::google::cloud::testing_util::SpanNamed;
 using ::google::cloud::testing_util::SpanWithParent;
 using ::testing::AllOf;
+using ::testing::ByMove;
 using ::testing::Contains;
+using ::testing::Return;
 
 pubsub::Subscription TestSubscription() {
   return pubsub::Subscription("test-project", "test-sub");
 }
 
 std::shared_ptr<MessageCallback> MakeTestMessageCallback(
-    std::shared_ptr<MessageCallback> mock) {
+    std::shared_ptr<MessageCallback> mock, Options opts) {
   return MakeTracingMessageCallback(
       std::move(mock),
-      Options{}.set<pubsub::SubscriptionOption>(TestSubscription()));
+      opts.set<pubsub::SubscriptionOption>(TestSubscription()));
 }
 
 TEST(TracingMessageCallback, UserCallback) {
   namespace sc = opentelemetry::trace::SemanticConventions;
   auto span_catcher = InstallSpanCatcher();
   auto mock = std::make_shared<pubsub_testing::MockMessageCallback>();
   EXPECT_CALL(*mock, user_callback).Times(1);
-  auto message_callback = MakeTestMessageCallback(std::move(mock));
-
+  auto message_callback = MakeTestMessageCallback(std::move(mock), Options{});
+  auto mock_ack_handler =
+      std::make_unique<pubsub_testing::MockExactlyOnceAckHandlerImpl>();
   auto span = internal::MakeSpan("my-sub subscribe");
-  MessageCallback::MessageAndHandler m{
-      pubsub::MessageBuilder().Build(),
-      std::make_unique<pubsub_testing::MockExactlyOnceAckHandlerImpl>(),
-      "ack-id",
-      {span}};
+  MessageCallback::MessageAndHandler m{pubsub::MessageBuilder().Build(),
+                                       std::move(mock_ack_handler),
+                                       "ack-id",
+                                       {span}};
+  message_callback->user_callback(std::move(m));
+  span->End();
+
+  auto spans = span_catcher->GetSpans();
+  EXPECT_THAT(
+      spans, Contains(AllOf(SpanHasInstrumentationScope(), SpanKindIsInternal(),
+                            SpanNamed("test-sub process"),
+                            SpanHasAttributes(OTelAttribute<std::string>(
+                                sc::kMessagingSystem, "gcp_pubsub")),
+                            SpanWithParent(span))));
+}
+
+TEST(TracingMessageCallback, AddTracingAckHandler) {
+  namespace sc = opentelemetry::trace::SemanticConventions;
+  auto span_catcher = InstallSpanCatcher();
+  auto mock_handler =
+      std::make_unique<pubsub_testing::MockExactlyOnceAckHandlerImpl>();
+  EXPECT_CALL(*mock_handler, ack_id).WillRepeatedly(Return("ack-id"));
+  EXPECT_CALL(*mock_handler, subscription).WillOnce(Return(TestSubscription()));
+  EXPECT_CALL(*mock_handler, delivery_attempt).WillOnce(Return(0));
+  EXPECT_CALL(*mock_handler, ack())
+      .WillOnce(Return(ByMove(make_ready_future(Status{}))));
```

Comment:
It might not be so useful, but we could check that `ThereIsAnActiveSpan()`.

It is not so useful because spans other than the `test-sub ack` span may be active. But maybe it is better than nothing?

Reply:
Added it, but I think it is a no-op.

Reply:
I don't know a better way to check. I thought about grabbing the spans here, like
```
  EXPECT_CALL(*mock_handler, ack()).WillOnce([&]() {
    ThereIsAnActiveSpan();
    auto spans = span_catcher->GetSpans();
    EXPECT_THAT(spans, testing::IsEmpty());
    return make_ready_future(Status{});
  });
```
(this test passes, when ideally it would fail and we could do stuff with the spans)

But the spans won't get collected since it doesn't end the span until the mock completes.